tokio-util/src/context.rs

```diff
@@ -0,0 +1,83 @@
+//! This module introduces `TokioContext`, which enables a simple way of using
+//! the tokio runtime with any other executor.
+
+use pin_project_lite::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::runtime::Handle;
+
+pin_project! {
+    /// TokioContext allows connecting a custom executor with the tokio runtime.
+    ///
+    /// It contains a `Handle` to the runtime. A handle to the runtime can be
+    /// obtain by calling the `Runtime::handle()` method.
+    pub struct TokioContext<F> {
+        #[pin]
+        inner: F,
+        handle: Handle,
+    }
+}
+
+impl<F: Future> Future for TokioContext<F> {
+    type Output = F::Output;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let me = self.project();
+        let handle = me.handle;
+        let fut = me.inner;
+
+        handle.enter(|| fut.poll(cx))
+    }
+}
+/// Trait extension that simplifies bundling a `Handle` with a `Future`.
+pub trait HandleExt: Into<Handle> + Clone {
```

Comment:
I don't think we actually need these extra bounds?

Reply:
I need to _convert_ HandleExt into a `TokioContext<F>`
```rust
    fn wrap<F>(&self, fut: F) -> TokioContext<F>
    where
        F: Future,
    {
        TokioContext {
            inner: fut,
            handle: (*self).clone().into(),
        }
    }
```
Is there a better alternative?

Reply:
Yeah, actually now that I think about it we want `Clone + ?Sized` right? that should allow us to implement this trait on a `&Handle`. I think.

Reply:
But we still need a way to convert the trait into a struct inside the trait definiton. Not sure how to avoid `Into<Handle>`

Reply:
You can avoid it by not implementing it in the trait, but in the impl block instead.

Reply:
@Darksonn do you mean this:
```diff
diff --git a/tokio-util/src/context.rs b/tokio-util/src/context.rs
index 9b878558..15615c80 100644
--- a/tokio-util/src/context.rs
+++ b/tokio-util/src/context.rs
@@ -34,7 +34,7 @@ impl<F: Future> Future for TokioContext<F> {
 }
 
 /// Trait extension that simplifies bundling a `Handle` with a `Future`.
-pub trait HandleExt: Into<Handle> + Clone {
+pub trait HandleExt {
     /// Convenience method that takes a Future and returns a `TokioContext`.
     ///
     /// # Example: calling Tokio Runtime from a custom ThreadPool
@@ -61,15 +61,15 @@ pub trait HandleExt: Into<Handle> + Clone {
     ///         .wrap(async { delay_for(Duration::from_millis(2)).await }),
     /// );
     ///```
-    fn wrap<F>(&self, fut: F) -> TokioContext<F>
-    where
-        F: Future,
+    fn wrap<F: Future>(&self, fut: F) -> TokioContext<F>;
+}
+
+impl HandleExt for Handle {
+    fn wrap<F: Future>(&self, fut: F) -> TokioContext<F>
     {
         TokioContext {
             inner: fut,
-            handle: self.clone().into(),
+            handle: self.clone(),
         }
     }
 }
-
-impl HandleExt for Handle {}
```

Reply:
Yes.